Add option to ignore suffix of gmx executables

theavey opened the pull request:
In response to my very old issue, this fixes #124.

This adds an option in the config file called append_suffix. If it is set to 'no', the suffix of the gmx executables (such as "_mpi" or "_d") will be ignored when making the tool names. It's not ideal to not differentiate them, but it can make code much more consistent across computers.

The largest change here is that in order to properly test this, the config file has to be changed before gromacs is imported. pytest makes all imports relative to the highest package it can find. Therefore, when `tests` was nested in the `gromacs` folder, the tests would be imported as `gromacs.tests.test_...`. This means that the `gromacs.__init__.py` is run before even running any of the pytest setup functions. By moving the tests folder out of the gromacs package, the config file can be edited before the `gromacs.__init__.py` is ever called.

Because of this change, I've tried to refactor all the references to `gromacs.tests` or `gromacs/tests` that I could find. 

This PR another item to the travis testing matrix that tests this option. Using two new command line options to pytest, a link to the gmx executable called 'gmx_mpi' is put into the home folder and written in the tools section of the config file; and the config file is written to ignore the suffix of the gmx executables found.

theavey commented:
It seems that I was overly ambitious trying to make what I thought were trivial changes to the ConfigParser and some uses of it.

Seeing as you seem to want to continue Python 2 support, I can roll those changes back.

theavey commented on tests/test_config.py:
```diff
@@ -14,7 +14,13 @@
 
-def get_GMXRC():
+
+@pytest.fixture
```
Made GMXRC a fixture

theavey commented on tests/test_config.py:
```diff
@@ -26,7 +32,7 @@ def get_GMXRC():
     GMXRC = os.path.join(bindir, 'GMXRC')
     if not os.path.exists(GMXRC):
-        raise IOError(errno.EEXIST, "Could not find Gromacs setup file", GMXRC)
+        raise IOError(errno.ENOENT, "Could not find Gromacs setup file", GMXRC)
```
This now raises the error that the file does not exist instead of that it does exist (it would be returned if it did exist).

orbeckst replied:
oops. thanks for catching

richardjgowers commented:
@theavey what's the problem you're running into with Py2?  I took a quick look and it looks like the `super()` call isn't happy, but I don't quite get why, it looks fine to me

theavey commented:
So in Python 2, ConfigParser (or SafeConfigParser) are old-style classes (they don't inherit from `object`), so they can't be imported that way. If they were new-style classes, that syntax is correct.

It seems that one option is to add a second parent of `object`. I've done this and a couple other fixes for Python 2 in a different branch. If the tests pass, I'll pull those changes into my main branch so they will be reflected here.

theavey commented:
If the reason for moving the tests folder wasn't clear, I just wrote up a [stackoverflow answer](https://stackoverflow.com/a/54640587/3961920) on a very related topic that might explain it more clearly.

theavey commented:
Hi @orbeckst, this PR is not critical, but if any changes need to be made, sooner would be better for me. This is just fresh in my mind now, and it would be easier to work on now instead of having to get back into this later. Thanks!

richardjgowers commented on .travis.yml:
```diff
@@ -51,6 +54,9 @@ install:
 script:
   - cd ${TRAVIS_BUILD_DIR}
   - if [[ $TRAVIS_OS_NAME == 'osx' ]]; then ulimit -S -n 2048; fi
+  - if [[ $SUFFIX_TEST == 'true' ]]; then
```
this if is messy because the extra test config is defined in 3? different places across the yaml.  Could it work so that the extended `${setup_cmd}` is defined in the `matrix:` section only?  Ideally something like `setup_cmd = ${setup_cmd} + '--no-append-suffix'`

theavey replied:
Makes sense. I'll clean it up.

richardjgowers commented on gromacs/config.py:
```diff
@@ -234,9 +234,22 @@
+    _cf_getbool = ConfigParser.getboolean
+    _UNSET = object()
+
+    def _getboolean(self, section, option, fallback=_UNSET, **kwargs):
```
this deserves a comment or two..

orbeckst commented:
@theavey sorry, I hadn't seen the PR – just very busy at the moment, so wasn't checking my GitHub feed. Generally speaking, if you need a reaction include @orbeckst so that I get an email notification.

@richardjgowers is already looking at the PR (THANK YOU Richard!!) and I won't really have time until the end of the week so whatever @richardjgowers decides is fine with me.

orbeckst commented:
And yes, I'd like to keep Python 2.7 compatibility – too much old code that uses it.

orbeckst commented:
@theavey sorry for the slow responses from me. Could you please resolve the conflicts that are flagged here?

orbeckst commented:
Please ping me or @richardjgowers when we should review again. Thanks!

theavey commented:
Hi @orbeckst,

The conflicts were just that I had moved the tests.

I've added more comments on `ConfigParser.getboolean`. The main change here is that I backported the `fallback` kwarg for Python 2.

The mode changes were inadvertent. I think they must have just been 644 when I cloned the repository. I am not sure how else they would have been changed.

I've added to CHANGES and AUTHORS.

orbeckst commented:
@theavey many thanks for your contribution!

theavey commented:
Of course! Thank you for the package

orbeckst commented:
P.S.: We wanted to write a short JOSS paper #147 – if you're interested in participating, add yourself to the issue and add to the PR #158.

(Admittedly, I have been occupied with millions of other things and it's a bit slow going but it's going to happen eventually and anyone who contributed to the package has a chance to be on the paper.)
